mdsuite/transformations/map_molecules.py

```diff
@@ -444,20 +516,42 @@ def _build_indices_dict(self, indices: List[int], species: List[str]) -> dict:
         for i, item in enumerate(species):
             if i == 0:
                 indices_dict[item] = np.sort(
-                    list(filter(lambda x: x < lengths[i], indices))
+                    np.array(list(filter(lambda x: x < lengths[i], indices)))
                 ).tolist()
             else:
                 greater_array = list(filter(lambda x: x >= lengths[i - 1], indices))
                 constrained_array = list(filter(lambda x: x < lengths[i], greater_array))
                 indices_dict[item] = np.sort(
-                    np.array(constrained_array) - (lengths[i - 1] - 1)
+                    np.array(constrained_array) - (lengths[i - 1])
                 ).tolist()
 
         return indices_dict
 
     def run_transformation(self, molecules: dict):
         """
         Perform the transformation.
+
+        Parameters
+        ----------
+        molecules : dict
+                Molecule dictionary to use as reference. The reference component is the
+                most critical part. One can either use a smiles string or a reference
+                dict as demonstrated below.
+
+                e.g, the input for EMIM-PF6 ionic liquid would be:
+
+                .. code-block::
+
+                   {'emim': {'smiles': 'CCN1C=C[N+](+C1)C', 'amount': 20},
+                   'PF6': {'smiles': 'F[P-](F)(F)(F)(F)F', 'amount': 20}}
```

Comment:
can this be a dataclass? If not for the interface, then at least for the internal calculations?

Reply:
Yes I will work with the comment made previously to try this out.

Reply:
Just for reference, @PythonFZ and @christophlohrmann  this is what I will make a data class that user can input. The other one I can make a data-class but it is actually just used as a test assertion.

Reply:
I think the code would benefit for both being dataclasses, but yes. This one is a bit tricky, because it should be two dataclasses and I don't know what the best interfache would be here. @christophlohrmann Would you request the user to pass two dataclasses? I think I'd go for `{"emim": mds.Molecule(smiles='...', amount='...')}` ?

Reply:
I can make the test one data class but it is used because the output of the smiles decomposition is also a dict returned by the other part of the code which is why I compare them directly. So if it were a data class I would either have to convert the data class into a dict and then compare it or compare it value-wise which would be a proliferation of code for the test so I don't hugely see the point there. Nonetheless, it can be done. Within the code making that whole part, which is used in total for about 2 lines, a data class is quite unnecessary. 

For the user interface one, I think that the Molecule should the data-class, can also use different terminology like ParticleGroup or something to be more generic and then when the user passes it here I would do as above, instead of passing a dict for each molecule you pass a dataclass. I don't  think it is odd that they would have to pass two data classes when they have to currently pass two dicts which is what is being replaced.

Reply:
In terms of naming some ideas could be nice to throw around, it can be generically used for groups of particles but currently is only used for molecules. Perhaps a generic naming choice could be bad, or just good in terms of looking ahead. I can see relatively soon the extension to polymer.

Reply:
> I think the code would benefit for both being dataclasses, but yes. This one is a bit tricky, because it should be two dataclasses and I don't know what the best interfache would be here. @christophlohrmann Would you request the user to pass two dataclasses? I think I'd go for `{"emim": mds.Molecule(smiles='...', amount='...')}` ?

we could also go for putting the name into the molecule class. then the user passes a list of objects.

Reply:
> In terms of naming some ideas could be nice to throw around, it can be generically used for groups of particles but currently is only used for molecules. Perhaps a generic naming choice could be bad, or just good in terms of looking ahead. I can see relatively soon the extension to polymer.

I like Molecule. A Polymer is also a large molecule.

Reply:
An interesting though, perhaps for the other PR but is relevant here, we could have some method for adding a molecule to an experiment directly. So the user defines their Molecule dict which I have written now, and instead of running mol mapping with this dict, there is a more arbitrary way of adding a group to an experiment. In this case, mapping molecule would act exactly like a normal calculator as it would just read the molecules attached to the experiment.

Reply:
Resolved in a0330c4 and 675ce61